### PR TITLE
sys: add evtimer WIP

### PR DESCRIPTION
**Pull request by kaspar030**
I have this timer module lying around. It's design goals are:
- use less state per timer
- don't waste memory for practically useless precision

I ended up with using a single pointer and a 32bit value per timer, making it possible to have a timers that only use 8 bytes on a 32bit platform.

### CODE REVIEW

**Comment by miri64** [comment]
@kaspar030 what about https://github.com/RIOT-OS/RIOT/projects/4?

**Comment by miri64** [comment]
Sorry wrong link. Corrected in edit.

File: sys/evtimer/evtimer.c [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    }
+}
+
+void evtimer_add(evtimer_t *evtimer, evtimer_event_t *event)
+{
+    unsigned state = irq_disable();
+    DEBUG("evtimer_add(): adding event with offset %"PRIu32"\n", event->offset);
+
+    _update_head_offset(evtimer);
+    _add_event_to_list((evtimer_event_t*)&evtimer->events, event);
```
Comment (miri64) [inline sys/evtimer/evtimer.c]: Why the cast and why the address operator here?

File: sys/evtimer/evtimer.c [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+void evtimer_init(evtimer_t *evtimer, void(*handler)(void*))
+{
+    evtimer->timer.callback = handler;
+    evtimer->timer.arg = (void*) evtimer;
+    evtimer->events = NULL;
+}
+
+#if ENABLE_DEBUG == 1
```
Comment (miri64) [inline sys/evtimer/evtimer.c]: This requires any external caller to edit this file internally (which also causes printing of the messages above). The compile should optimize it away if it isn't called. Alternatively, put this function in another file (if you don't trust the compiler optimization).

File: sys/include/div.h [hunk 3]
```diff
@@ -45,6 +45,33 @@ static inline uint64_t div_u64_by_15625(uint64_t val)
 }
 
 /**
+ * @brief Integer divide val by 125
+ *
+ * This function can be used to convert uint64_t microsecond times (or
+ * intervals) to miliseconds and store them in uint32_t variables, with up to
+ * ~50 days worth of miliseconds ((2**32*1000) -1).
```
Comment (miri64) [inline sys/include/div.h]: milliseconds

File: sys/evtimer/evtimer.c [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    DEBUG("evtimer_msg_handler()\n");
+
+    evtimer_t *evtimer = (evtimer_t*) arg;
+
+    evtimer_event_t *event = evtimer->events;
+    event->offset = 0;
+
+    evtimer_msg_event_t *mevent;
+    while ((mevent = (evtimer_msg_event_t *)_get_next(evtimer))) {
+        msg_send_int(&mevent->msg, mevent->msg.sender_pid);
```
Comment (miri64) [inline sys/evtimer/evtimer.c]: Send to `sender_pid`? Also: isn't this value usually uninitialized by the user?

Easy fix: see [here](https://github.com/kaspar030/RIOT/pull/19/files#diff-86e356dc5223dce7b4926e951244cfc0R77)

**Comment by zhuoshuguo** [comment]
@kaspar030 @miri64 could you provide a simple `evtimer` test example? or something like a local test branch I can follow? By playing with it, I think I can follow more quickly. :-)

**Comment by zhuoshuguo** [comment]
By playing with it, I found a possible bug: 
I simply couldn't add the first event timeout:
```
    evtimer_t evtimer; 
    evtimer_msg_event_t msg_event;

    ......
    msg_event.event.next= NULL;
    msg_event.event.offset = 2000;
    msg_event.msg.type = GNRC_EVTIMEOUT_TYPE;
    msg_event.msg.content.ptr = &msg_event;
    msg_event.msg.sender_pid = gnrc_netdev2->pid;

    evtimer_init(&evtimer, evtimer_msg_handler);
    evtimer_add(&evtimer, (evtimer_event_t *)&msg_event);

```
Then, I checked and found that `_add_event_to_list()` (also `_del_event_from_list()`) seems to have malfunction, please see below:

File: sys/evtimer/evtimer.c [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        next_entry->offset -= event->offset;
+    }
+    event->offset -= delta_sum;
+
+    list->next = event;
+}
+
+static void _del_event_from_list(evtimer_event_t *list, evtimer_event_t *event)
+{
+    while (list->next) {
```
Comment (zhuoshuguo) [inline sys/evtimer/evtimer.c]: Similar issue here.

File: sys/include/evtimer.h [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+typedef struct {
+    xtimer_t timer;
+    evtimer_event_t *events;
+} evtimer_t;
+
+struct evtimer_event {
+    evtimer_event_t *next;
+    uint32_t offset;
+};
```
Comment (zhuoshuguo) [inline sys/include/evtimer.h]: Do you think we can add a `expired` bool value into the `evtimer_event_t` structure, which I think should be commonly used by users? :-)
@miri64 If not, I think I will add it (`expired`) into my `gnrc_mac` timeout module which will probably be the wrapper of evtimer module.

Comment (miri64) [inline sys/include/evtimer.h]: @zhuoshuguo it's more RAM saving to implement `expired` as a function if you actually need it (though usually it is more advisable to just react via callback to an expired event ;-)): You could check the event list if the event is still contained.

Comment (zhuoshuguo) [inline sys/include/evtimer.h]: @miri64 I think you are right~ :smile: Thanks!

**Comment by zhuoshuguo** [comment]
Add checking event running/scheduled functionality (see [add_check_event_sched](https://github.com/kaspar030/RIOT/pull/21/files)) which I think also enables checking event expired. Should be at least useful for me. :smile:

**Comment by kaspar030** [comment]
> Then, I checked and found that _add_event_to_list() (also _del_event_from_list()) seems to have malfunction, please see below and fix

Sorry, what was a too-quickly fix in the last commit. Both functions' logics required to get a pointer to pointer to the first list element in order to work, which I diffed out accidentaly. I've moved the pointer logic into the functions now.

**Comment by zhuoshuguo** [comment]
@kaspar030 Sorry for my impetuous action..

**Comment by kaspar030** [comment]
> @kaspar030 Sorry for my impetuous action..

Don't be sorry, thanks for fixing my mess!

**Comment by miri64** [comment]
@kaspar030 can you provide (and maybe even tests)? I need to use this soon-ish for NDP. Let's merge this with #6000 later on (especially since we need to overhaul IPC to use thread flags for my use-case anyways to make it useful for my context anyways :-/).

File: sys/include/div.h [hunk 7]
```diff
@@ -45,6 +45,33 @@ static inline uint64_t div_u64_by_15625(uint64_t val)
...
+ * Use e.g.,  ms = div_u64_by_125(microseconds >> 3)
+ *
+ * @pre val <= 536870911999 ((2**32 * 125) -1)
+ *
+ * @param[in]   val     dividend
+ * @return      (val / 125)
+ */
+static inline uint32_t div_u64_by_125(uint64_t val)
+{
+  /* a higher value would overflow the result type */
```
Comment (miri64) [inline sys/include/div.h]: Since when do we use two spaces?

**Comment by miri64** [comment]
Also, please consider https://github.com/kaspar030/RIOT/pull/22. This streamlines the message-based API a little bit.

**Comment by miri64** [comment]
Also doesn't build atm... Will see, if I can provide a fix.

**Comment by miri64** [comment]
Ah, the changes in #5608 were the problem.

**Comment by miri64** [comment]
I guess we postpone then.

**Comment by zhuoshuguo** [comment]
I will be also thankful if this PR will get merged soon, since my ```gnrc_mac```'s timeout module is also based on evtimer. :smile:

**Comment by emmanuelsearch** [comment]
@kaspar030 any news on this? What should be done in #6554 at this point?

**Comment by zhuoshuguo** [comment]
If this PR is still WIP, I suggest we push #6554 forward without it (evtimer). 

While when this PR get merged someday in the future, I will push [#5949](https://github.com/RIOT-OS/RIOT/pull/5949) to merge, and reflect evtimer in #6554.

**Comment by miri64** [comment]
PING?

**Comment by miri64** [comment]
Adapted in #6767

**Comment by kaspar030** [comment]
Closed in favour of 6767. Thanks for taking over!

**Pull request closed without merging**